### PR TITLE
[xray-exporter] Add ability to publish attachments for test executions

### PR DESCRIPTION
**Pull request by uarlouski**

### CODE REVIEW

File: vividus-facade-jira/src/main/java/org/vividus/jira/JiraFacade.java [hunk 1]
```diff
@@ -83,6 +88,16 @@ public JiraEntity getIssue(String issueKey) throws IOException, JiraConfiguratio
...
     }
 
+    public void addAttachment(String issueKey, String attachmentName, byte[] attachmentData)
+            throws IOException, JiraConfigurationException
+    {
+        HttpEntity attachmentEntity = MultipartEntityBuilder.create()
+                .addBinaryBody("file", attachmentData, ContentType.MULTIPART_FORM_DATA, attachmentName)
+                .build();
+        jiraClientProvider.getByIssueKey(issueKey).executePost(ISSUE_ENDPOINT + issueKey + "/attachments",
+                List.of(new BasicHeader("X-Atlassian-Token", "no-check")), attachmentEntity);
```
Comment (valfirst) [inline vividus-facade-jira/src/main/java/org/vividus/jira/JiraFacade.java:98]: what is this header for?

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-facade-jira/src/main/java/org/vividus/jira/JiraFacade.java (hunk 1 above)
Comment (uarlouski) [reply vividus-facade-jira/src/main/java/org/vividus/jira/JiraFacade.java:98]: This header is mentioned in documentation [here](https://developer.atlassian.com/cloud/jira/platform/rest/v3/api-group-issue-attachments/#api-rest-api-3-issue-issueidorkey-attachments-post) and if we do not provide it the endpoint throws an error, I lost the message but this header disables some verifications

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/facade/XrayFacade.java [hunk 2]
```diff
@@ -108,15 +115,59 @@ public void importTestExecution(TestExecution testExecution) throws IOException,
...
+
+    public void addAttachments(String testExecutionKey, List<Path> attachments)
+            throws IOException, JiraConfigurationException
+    {
+        for (Path attachmentPath : attachments)
+        {
+            try
+            {
+                String name = FilenameUtils.getName(attachmentPath.toString());
+                if (name.isEmpty() || !Files.isDirectory(attachmentPath) && !Files.isRegularFile(attachmentPath))
```
[the remark this answers is not in the document]
Comment (valfirst) [reply vividus-to-xray-exporter/src/main/java/org/vividus/xray/facade/XrayFacade.java:139]: how can something be neither directory nor file? if it doesn't exist, we should throw explicit error

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/configuration/validator/XrayExporterOptionsValidator.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        XrayExporterOptions options = (XrayExporterOptions) target;
+        List<Path> attachments = options.getTestExecutionAttachments();
+
+        attachments.forEach(attachment ->
+        {
+            if (!Files.exists(attachment))
+            {
+                errors.rejectValue(TEST_EXECUTION_ATTACHMENTS_FIELD, StringUtils.EMPTY,
+                        "Attachment file by path " + attachment + " does not exist");
```
Comment (valfirst) [inline vividus-to-xray-exporter/src/main/java/org/vividus/xray/configuration/validator/XrayExporterOptionsValidator.java:54]:
```suggestion
                        "The attachment file at path " + attachment + " does not exist");
```

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/configuration/validator/XrayExporterOptionsValidator.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                return;
+            }
+
+            try
+            {
+                if (Files.isDirectory(attachment) && Files.list(attachment).count() == 0)
+                {
+                    errors.rejectValue(TEST_EXECUTION_ATTACHMENTS_FIELD, StringUtils.EMPTY,
+                            "The attachment foldedr by path " + attachment + " is empty");
```
Comment (valfirst) [inline vividus-to-xray-exporter/src/main/java/org/vividus/xray/configuration/validator/XrayExporterOptionsValidator.java:63]:
```suggestion
                            "The attachment folder at path " + attachment + " is empty");
```

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/facade/XrayFacade.java [hunk 5]
```diff
@@ -109,15 +118,63 @@ public void importTestExecution(TestExecution testExecution) throws IOException,
...
+                    name += ".zip";
+                    body = output.toByteArray();
+                }
+                else
+                {
+                    body = Files.readAllBytes(attachmentPath);
+                }
+
+                attachments.add(new Attachment(name, body));
+            }
```
Comment (valfirst) [inline vividus-to-xray-exporter/src/main/java/org/vividus/xray/facade/XrayFacade.java:164]: it makes sense to move attachments preparation to a separate method

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/facade/XrayFacade.java [hunk 6]
```diff
@@ -109,15 +118,63 @@ public void importTestExecution(TestExecution testExecution) throws IOException,
...
+
+            jiraFacade.addAttachments(testExecutionKey, attachments);
+
+            LOGGER.atInfo().addArgument(attachmentPaths)
+                           .addArgument(testExecutionKey)
+                           .log("Successfully attached {} files to test execution with key {}");
```
Comment (valfirst) [inline vividus-to-xray-exporter/src/main/java/org/vividus/xray/facade/XrayFacade.java:170]:
```suggestion
                           .log("Successfully attached files and folders at {} to test execution with key {}");
```

File: vividus-to-xray-exporter/src/main/java/org/vividus/xray/facade/XrayFacade.java [hunk 7]
```diff
@@ -109,15 +118,63 @@ public void importTestExecution(TestExecution testExecution) throws IOException,
...
+        }
+        catch (IOException | ZipException | JiraConfigurationException e)
+        {
+            LOGGER.atError().setCause(e)
+                            .addArgument(attachmentPaths)
+                            .addArgument(testExecutionKey)
+                            .log("Failed to attach {} files to test execution with key {}");
```
Comment (valfirst) [inline vividus-to-xray-exporter/src/main/java/org/vividus/xray/facade/XrayFacade.java:177]:
```suggestion
                            .log("Failed to attach files and folders at {} to test execution with key {}");
```

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-to-xray-exporter/src/test/java/org/vividus/xray/facade/XrayFacadeTests.java [hunk 8]
```diff
@@ -228,12 +279,65 @@ void shouldCreateTestExecution() throws JiraConfigurationException, IOException
...
+        assertThat(logger.getLoggingEvents(), is(List.of(
+            info(UPDATE_TEST_EXECUTION_LOG, ISSUE_KEY, TEST_EXECUTION_REQUEST),
+            error(thrown, "Failed to attach files and folders at {} to test execution with key {}", List.of(directory),
+                    ISSUE_KEY),
+            info(TEST_EXECUTION_UPDATED_LOG, ISSUE_KEY)
+        )));
+    }
+
+    @Test
+    void shouldNotAttachAnythinIfAttachmentsAreEmpty() throws JiraConfigurationException, IOException
```
Comment (valfirst) [inline vividus-to-xray-exporter/src/test/java/org/vividus/xray/facade/XrayFacadeTests.java:325]:
```suggestion
    void shouldNotAttachAnythingIfAttachmentsAreEmpty() throws JiraConfigurationException, IOException
```

**Review by valfirst — APPROVED** [review]

**Pull request merged**